[ch16760] Revisit Apple Pay Order Creation

wvega opened the pull request:
# Summary

This PR updates `SV_WC_Payment_Gateway_Apple_Pay_Orders::create_order()` to use `\WC_Checkout` methods so that creating orders for Apple Pay is as similar as possible to creating an order from the checkout page.

Particularly, we wanted that:

- all filters and actions that fire during regular checkout also fire when an Apple Pay order is created
- product options configured through the TM Global Extra Product Options plugin are included for product line items on Apple Pay orders

### Stories

- [CH 16760](https://app.clubhouse.io/skyverge/story/16760/revisit-apple-pay-order-creation)

## Details

Using Apple Pay now requires WooCommerce 3.2. A notice will be shown on older versions.

I set the version to 5.5.1, but maybe that's not the best choice since we are dropping Apple Pay support for WC < 3.2?

This PR doesn't include a commit to update the namespace on purpose to facilitate testing with the version of Authorize.Net in the master branch on [wc-plugins](https://github.com/skyverge/wc-plugins).

## UI

Admin notice shown if Apply Pay is enabled on WooCommerce 3.1.x or older: [screenshot](https://share.getcloudapp.com/xQuvE8Rp)

## QA

### Setup

**Authorize.Net**

Use the version of Authorize.Net in the master branch on [wc-plugins](https://github.com/skyverge/wc-plugins):
- Require version `dev-ch16760/revisit-apple-pay-order-creation#a23e6a8` of `skyverge/wc-plugin-framework` in `composer.json`
- Run `composer update`

**Test Device**

- You need a device where Apple Pay is supported
- You can use [Sandbox Tester account](https://developer.apple.com/apple-pay/sandbox-testing/) if not using your regular account.
- If you use a Sandbox Tester account, you can add the following [Test Card](https://developer.apple.com/apple-pay/sandbox-testing/) to your Wallet:
    * Number: 4761 1200 1000 0492
    * Expiration Date: 11/2022
    * CVV: 533
- Make sure to have a Billing Address, Shipping Address, Email Address, and Phone Number configured in Settings > Wallet & Apple Pay 

**TM Extra Product Options**

- Install and activate TM Extra Product Options plugin
- Go to Products > TM Global Extra Product Options
- Add a Global Form
- Enter a name for the global form
- Use the Add a Section button to add a section
    - Add a Radio Buttons field and edit its settings
        - Scroll to the bottom of the General Options tab and add A and B options with no price
- Add another section and move it to the bottom of the form
    - Add a Checkboxes field and edit its settings
        - Add options A1, A2, A3 with a different price for each
        - Go to the Conditional Logic tab and add a rule to show this field the value in the Radio Buttons field is A
    - Add another Checkboxes field and edit its settings
        - Add options B1, B2, B3 with a different price for each
        - Go to the Conditional Logic tab and add a rule to show this field the value in the Radio Buttons field is B
- Publish the global form

**WooCommerce**

- Configure the Test Enviroment for Authorize.Net Credit Card with Tokenization enabled.
- Disable Test Mode in Apple Pay settings so that payments are processed by the payment gateway
- Enable taxes and add 5% tax rate for any country code, state code, postcode or city
- Enable coupons and create a $10 coupon with no usage restriction
- Edit a Product and go to the TM Extra Product Options tab (below Attributes)
    - Add Radio Buttons field with two options and a price for each option
    - Switch to the Settings subtab
    - Check the option for the global form on the **Include additional Global forms** setting
    - Update the product

### Cases

Using your test device:

#### Create an order from the product page

1. Navigate to the modified product
1. Click the Buy with Pay button and pay
    - [x] The order is received successfully
    - [x] The order includes a line item for the modified product
    - [x] The order incldues a shipping item
    - [x] The order includes a tax item

#### Create an order from the cart page

1. Navigate to the modified product
1. Select different values for the extra product options so that the options amount is a non zero value
1. Change the quantity to a number other than 1
1. Add the product to the cart
1. Go to the cart page
1. Apply the test coupon
1. Click the Pay and pay
    - [x] The order is received successfully
    - [x] The order includes a line item for the modified product
    - [x] Selected extra product options match the values selected on the product page
    - [x] The order incldues a discount item
    - [x] The order incldues a shipping item
    - [x] The order includes a tax item
    - [x] The quantity of the produce matches the selected product
    - [x] The totals and subtotals were calculated correctly

#### Create an order from the checkout page

1. Navigate to the modified product
1. Select different values for the extra product options so that the options amount is a non zero value
1. Change the quantity to a number other than 1
1. Add the product to the cart
1. Go to the checkout page
1. Apply the test coupon
1. Click the Pay and pay
    - [x] The order is received successfully
    - [x] The order includes a line item for the modified product
    - [x] Selected extra product options match the values selected on the product page
    - [x] The order incldues a discount item
    - [x] The order incldues a shipping item
    - [x] The order includes a tax item
    - [x] The quantity of the produce matches the selected product
    - [x] The totals and subtotals were calculated correctly

#### Apple Pay not supported on WooCommerce < 3.2

1. Test using WooCommerce 3.1.2 (or ⚠️ add `define( 'WC_VERSION', '3.1.2' );` to your `wp-config.php`)
    - [x] An admin notice shows up indicating that Apple Pay requires WooCommerce 3.2
    - [x] Apple Pay settings are not available on WooCommerce > Settings > Payments

## Before merge

- [ ] I have confirmed these changes in each supported minor WooCommerce version

ChaseWiseman commented on woocommerce/changelog.txt:
```diff
@@ -1,5 +1,8 @@
 *** SkyVerge WooCommerce Plugin Framework Changelog ***
 
+2019.nn.nn - version 5.5.1
+ * Fix - Refactor Apple Pay order creation to support the same filters and actions that are fired during regular checkout
```
```suggestion
 * Tweak - Refactor Apple Pay order creation to support the same filters and actions that are fired during regular checkout
```

ChaseWiseman commented on woocommerce/payment-gateway/apple-pay/class-sv-wc-payment-gateway-apple-pay-orders.php:
```diff
@@ -49,188 +49,49 @@ class SV_WC_Payment_Gateway_Apple_Pay_Orders {
 
 		try {
 
 			wc_transaction_query( 'start' );
```
just checking - are these `wc_transaction_query()` calls still relevant? I don't see them used in core anymore.

wvega replied:
I think we can remove the calls to  `wc_transaction_query()` now. They made more sense when we were trying to save every item as it was created, but as you said, they are not used in the core order creation process anymore. I don't see them anywhere else in the latest version of the code either.

Update in f0ae221.

wvega commented:
I'm sorry for the delay on addressing feedback here. Thank you for updating the base branch and merging `5.5.1-dev`.

I changed the notice to appear on WC settings screens only in c8f0292.
